renderlib/gesture/gesture.cpp

```diff
@@ -0,0 +1,559 @@
+
+      // Record position of the pointer during the press event, we are going to use it to
+      // determine drag operations
+      button.pressedPosition = position;
+
+      // The state of modifiers are recorded when buttons are initially pressed. The state
+      // is retained for the duration of the click/drag. We do not update this state until the
+      // next button press event.
+      button.modifier = mods;
+    }
+
+    button.action = Gesture::Input::kPress;
+  } else if (action == Input::kRelease) {
+    // When button is released, record any drag distance
+    if (button.action != Gesture::Input::kNone) {
+      glm::vec2 origin = button.pressedPosition;
+      button.drag = position - origin;
+    }
+
+    button.action = Gesture::Input::kRelease;
+  }
+}
+
+void
+Gesture::Input::setPointerPosition(glm::vec2 position)
+{
+  cursorPos = position;
+
+  // Update each button action. Each button holding a kPress event becomes a
+  // kDrag event.
+  for (int mbIndex = 0; mbIndex < Input::kButtonsCount; ++mbIndex) {
+    Button& button = mbs[mbIndex];
+    if (button.action == Input::kNone || button.action == Input::kRelease) {
+      continue;
+    }
+
+    glm::vec2 origin = button.pressedPosition;
+    glm::vec2 drag = position - origin;
+    button.drag = drag;
+    bool anyMotion = drag != glm::vec2(0);
+
+    if (button.action == Gesture::Input::kPress && anyMotion) {
+      button.action = Gesture::Input::kDrag;
+
+      // If we hold the shift modifier we record if the initial drag is mostly
+      // horizontal or vertical. This information may be used by some tools.
+      if (button.modifier & Gesture::Input::kShift) {
+        button.dragConstraint = (abs(drag.x) > abs(drag.y) ? Gesture::Input::kHorizontal : Gesture::Input::kVertical);
+      } else {
+        button.dragConstraint = Gesture::Input::kUnconstrained;
+      }
+    }
+  }
+}
+
+namespace Pipeline {
+
+// First I may draw any GUI geometry that I want to be depth-composited with the
+// rest of the scene in viewport. This may be any supporting guide that needs to
+// appear for the duration of some action and reveal intersections against the
+// scene geometry.
+
+// Draw something "in the scene". This has a limitation that we assume there is a
+// single viewport.
+static void
+configure_3dDepthTested(SceneView& sceneView, Gesture::Graphics& graphics)
+{
+  auto& shader = graphics.shader;
+
+  glm::mat4 v(1.0);
+  sceneView.camera.getViewMatrix(v);
+  glm::mat4 p(1.0);
+  sceneView.camera.getProjMatrix(p);
+
+  glUniformMatrix4fv(shader->m_loc_proj, 1, GL_FALSE, glm::value_ptr(p * v));
+  check_gl("set proj matrix");
+
+  glEnable(GL_DEPTH_TEST);
+  check_gl("enable depth test");
+}
+
+// The second pass is still about 3d geometry, only this time I want it to be
+// drawn on top, without depth test. These two passes shares in common the same
+// projection matrix as the rest of the scene. 3d manipulators shown earlier are examples.
```

Comment:
Appreciated these comments! Not sure if they were from the blog but it was helpful in understanding how the Pipeline worked.